src/hyperlight_host/tests/integration_test.rs

```diff
@@ -14,19 +14,236 @@ See the License for the specific language governing permissions and
+    assert!(!interrupt_handle.dropped()); // not yet dropped
+
+    // kill vm after 1 second
+    let thread = thread::spawn(move || {
+        thread::sleep(Duration::from_secs(1));
+        assert!(interrupt_handle.kill());
+        barrier2.wait(); // wait here until main thread has returned from the interrupted guest call
+        barrier2.wait(); // wait here until main thread has dropped the sandbox
+        assert!(interrupt_handle.dropped());
+    });
+
+    let res = sbox1
+        .call_guest_function_by_name::<i32>("Spin", ())
+        .unwrap_err();
+    assert!(matches!(res, HyperlightError::ExecutionCanceledByHost()));
+
+    barrier.wait();
+    // Make sure we can still call guest functions after the VM was interrupted
+    sbox1
+        .call_guest_function_by_name::<String>("Echo", "hello".to_string())
+        .unwrap();
+
+    // drop vm to make sure other thread can detect it
+    drop(sbox1);
+    barrier.wait();
+    thread.join().expect("Thread should finish");
+}
+
+/// Makes sure interrupting a vm before the guest call has started also prevents the guest call from being executed
+#[test]
+fn interrupt_guest_call_in_advance() {
+    let mut sbox1: MultiUseSandbox = new_uninit_rust().unwrap().evolve(Noop::default()).unwrap();
+    let barrier = Arc::new(Barrier::new(2));
+    let barrier2 = barrier.clone();
+    let interrupt_handle = sbox1.interrupt_handle();
+    assert!(!interrupt_handle.dropped()); // not yet dropped
+
+    // kill vm before the guest call has started
+    let thread = thread::spawn(move || {
+        assert!(!interrupt_handle.kill()); // should return false since vcpu is not running yet
+        barrier2.wait();
+        barrier2.wait(); // wait here until main thread has dropped the sandbox
+        assert!(interrupt_handle.dropped());
+    });
+
+    barrier.wait(); // wait until `kill()` is called before starting the guest call
+    let res = sbox1
+        .call_guest_function_by_name::<i32>("Spin", ())
+        .unwrap_err();
+    assert!(matches!(res, HyperlightError::ExecutionCanceledByHost()));
+
+    // Make sure we can still call guest functions after the VM was interrupted
+    sbox1
+        .call_guest_function_by_name::<String>("Echo", "hello".to_string())
+        .unwrap();
+
+    // drop vm to make sure other thread can detect it
+    drop(sbox1);
+    barrier.wait();
+    thread.join().expect("Thread should finish");
+}
+
+/// Verifies that only the intended sandbox (`sbox2`) is interruptible,
+/// even when multiple sandboxes share the same thread.
+/// This test runs several interleaved iterations where `sbox2` is interrupted,
+/// and ensures that:
+/// - `sbox1` and `sbox3` are never affected by the interrupt.
+/// - `sbox2` either completes normally or fails with `ExecutionCanceledByHost`.
+///
+/// This test is not foolproof and may not catch
+/// all possible interleavings, but can hopefully increases confidence somewhat.
+#[test]
+fn interrupt_same_thread() {
```

Comment:
Would be cool to try:
(1) create sbox1 and sbox2 on thread_a
(2) run Spin guest function for sbox1 on thread_b + run a guest function that can conclude for sbox2 on thread_c.
(3) cancel execution for sbox1.

Will (3) kill sbox2 as well or does this work fine because we re-set the tid on `run`?